src/librustc/ty/mod.rs

```diff
@@ -767,6 +767,73 @@ impl ty::EarlyBoundRegion {
     }
 }
 
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub enum GenericParam {
+    Lifetime(RegionParameterDef),
+    Type(TypeParameterDef),
+}
+
+impl GenericParam {
+    pub fn index(&self) -> u32 {
+        match self {
+            GenericParam::Lifetime(lt) => lt.index,
+            GenericParam::Type(ty)     => ty.index,
+        }
+    }
+}
+
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub enum KindIndex {
+    Lifetime,
+    Type,
+}
+
+#[derive(Clone, Debug, RustcEncodable, RustcDecodable)]
+pub struct KindIndexed<L, T> {
```

Comment:
Do we really want this? Also I'd keep such a refactoring separate from AST/HIR IMO.

Reply:
Yeah, it's not used in very many places... I did it simply to avoid magic constants, but I was debating whether it was worth the extra declaration. I can separate it out regardless, though.